src/main/scala/com/snowplowanalytics/snowplow/rdbloader/Main.scala

```diff
@@ -12,71 +12,69 @@
-import interpreters.Interpreter
-import config.CliConfig
-import loaders.Common.{ load, discover }
+import com.snowplowanalytics.snowplow.rdbloader.dsl.{AWS, JDBC, Logging, RealWorld}
+import com.snowplowanalytics.snowplow.rdbloader.config.CliConfig
+import com.snowplowanalytics.snowplow.rdbloader.loaders.Common.{discover, load}
+import com.snowplowanalytics.snowplow.rdbloader.utils.{S3, SSH}
 
-/**
- * Application entry point
- */
-object Main {
+object Main extends IOApp {
   /**
    * If arguments or config is invalid exit with 1
    * and print errors to EMR stdout
    * If arguments and config are valid, but loading failed
    * print message to `track` bucket
    */
-  def main(argv: Array[String]): Unit = {
+  def run(argv: List[String]): IO[ExitCode] =
     CliConfig.parse(argv) match {
       case Valid(config) =>
-        val status = run(config)
-        sys.exit(status)
-      case Invalid(errors) =>
-        println("Configuration error")
-        errors.toList.foreach(error => println(error.message))
-        sys.exit(1)
-    }
-  }
+        RealWorld.initialize[IO](config).flatMap { dsls =>
+          import dsls._
 
-  /**
-   * Initialize interpreter from parsed configuration and
-   * run all IO actions through it. Should never throw exceptions
-   *
-   * @param config parsed configuration
-   * @return exit code status. 0 for success, 1 if anything went wrong
-   */
-  def run(config: CliConfig): Int = {
-    val interpreter = Interpreter.initialize(config)
+          val result = for {
+            discovery <- discover[IO](config)
+            jdbc = SSH.resource[IO](config.target.sshTunnel) *>
+              JDBC.interpreter[IO](config.target, config.dryRun)
+            _ <- LoaderAction(jdbc.use { implicit conn => load[IO](config, discovery).value })
+          } yield ()
 
-    val actions: Action[Int] = for {
-      data       <- discover(config).flatTap(db.Migration.perform(config.target.schema)).value
-      result     <- data match {
-        case Right(discovery) => load(config, discovery).value
-        case Left(LoaderError.StorageTargetError(message)) =>
-          val upadtedMessage = s"$message\n${interpreter.getLastCopyStatements}"
-          ActionE.liftError(LoaderError.StorageTargetError(upadtedMessage))
-        case Left(error) => ActionE.liftError(error)
-      }
-      message     = utils.Common.interpret(config, result)
-      _          <- LoaderA.track(message)
-      status     <- close(config.logKey, message)
-    } yield status
-
-    actions.foldMap(interpreter.run)
-  }
+          result
+            .value
+            .attempt
+            .map {      // TODO: write shorter; and figure out if unit test is possible
+              case Left(e) =>
+                e.printStackTrace(System.out)
+                (LoaderError.LoaderLocalError(e.getMessage): LoaderError).asLeft
+              case Right(e) => e
+            }
+            .flatMap(res => close[IO](config.logKey, res))
```

Comment:
Can't we use `Bracket` or `Resource` instead of manually calling `close()` ?

Reply:
I think you're right, `Resource` would work out. I just didn't consider it probably because `close` doesn't anything super important, but will check it out in R34.